### ISSUE TITLE
HubSpot CRM issue

### ISSUE THREAD

**Issue by Uthoocha** [microsoft/PowerPlatformConnectors#2309]
Api Name - shared_hubspotcrm

Bug description - Error on connection - Couldn’t complete the connection
Insufficient scopes were provided. Please contact the app developer.

Is this a security bug? (Y/N) I think so?

What is the severity of this bug? Can't use the connector.
This problem maybe my fault for no understanding Github read me instructions. Do you first configure an instant cloud flow using the 'When a HTTP request is received' trigger???

**Comment by troystaylor** [comment]
Hello,
Have you configured your HubSpot OAuth app to include the following scopes?:

crm.objects.contacts.read tickets crm.objects.marketing_events.read content reports actions timeline oauth forms files crm.objects.marketing_events.write crm.schemas.custom.read crm.objects.custom.read crm.objects.feedback_submissions.read crm.objects.custom.write

**Comment by Uthoocha** [comment]
Yes I did, but I think the issue here is to do with the new Hubspot API change that happened Nov 2022. Forcing Non Dev account users to create these private app thingies, with Access Tokens as opposed to API keys. 

Anyway, I've managed to get this working, however, not sure how to share this with the community. So I will explain it here and see if this is of value to you or anyone else.

I watched your video on creating custom connections in Power Automate.
I started to create a custom connector.
I opened the Swagger Editor and pasted the content of the file apiDefinition.swagger.json
_PowerPlatformConnectors/tree/dev/independent-publisher-connectors/HubSpot%20CRM)/apiDefinition.swagger.json_
I then went back to the Security Tab (2.Security)
Selected API Key as Authentication type.
Parameter label * = API Key
Parameter name * = Authorization
Parameter location * = Header

Went to Test tab (5, Test) , clicked + New Connection
And then entered my Private App Access token like so ' Bearer ***-***-********-****-****-****-************ '

First attempt got a 403 error due to missing Scope requirements. (crm.objects.companies.read - crm.schemas.companies.read)
So I added those scopes (This maybe why the actual Plugin is showing the error I originally opened the ticket for)

Full list of scopes to make this work with Private App and Access Key.
Scopes: content, reports, actions, timeline, oauth, forms, files, tickets, crm.objects.contacts.read
crm.objects.marketing_events.read
crm.objects.marketing_events.write
crm.schemas.custom.read
crm.objects.custom.read
crm.objects.custom.write
crm.objects.feedback_submissions.read
crm.objects.companies.read
crm.schemas.companies.read


I am sure you will be able to recreate what I did. I think I will keep with the Custom Connect if that is allowed by you, as I wan't to under how this all works because I think i got a little lucky here as I am not a Dev.

**Comment by troystaylor** [comment]
You got a little lucky and at some point, your custom connector will stop working if it uses an API key. There is no need to create the connector as custom if you are just testing, but you must use OAuth.

**Comment by marctre123** [comment]
@troystaylor is that now considered a known issue? We're also getting the same error Insufficient scopes were provided. Please contact the app developer. 
How do we go about installing the OAuth app for your connector that you mentioned? Do we need to create it ourselves or will it install automatically when we create a connection for your connector? If it is the later, then it doesn't happen because we get the error  when the App Install url is requested.

**Comment by troystaylor** [comment]
You don't install my OAuth app, you have to create your own in a HubSpot developer account. Then add the scopes to your OAuth app.

**Comment by marctre123** [comment]
@troystaylor apologies if I'm missing something obvious. 

I could have dozen of apps in my HubSpot account, so how is the connector going to know which one to use when all it asks is to choose an environment? If I was to create a custom connector, then I know I would have to provide the client ID and secret in the generic OAuth 2 configuration to identify the app, but in the case that I'm using your connector, wasn't that done on your side?

It would be helpful to have an updated readme file if the connector still works with the recent HubSpot API changes. Like @Uthoocha mentioned, API keys are deprecated and the readme still revolves around this approach.

**Comment by Uthoocha** [comment]
@marctre123 I just made a custom connector and changed the security settings to API, submitted the Access key from a Private App and done, connecter is working, I also changed a few bits so that I can look up contacts in HS using email instead of Contact ID. I am most likely wrong, but I think if Troy adds crm.objects.companies.read - crm.schemas.companies.read to the scopes of the community connector, his app will work again. I say this because as I went down my route, I got a 403 error saying these scopes were missing, I added them to my Private App and it works. Anyway, I don't know how this whole github thing works, so I tried to create a Fork or something and add the scopes to the community app, but I don't have write access, I also don't know how I test my fork. Any help, let me know.

**Comment by marctre123** [comment]
@Uthoocha I agree it seems like the connector is not passing the right scopes to install the app. I thought maybe adding all the possible scopes by hand would work, but instead it gives me a different error:
Couldn’t complete the connection
This account doesn't have access to some HubSpot features that are required by this app. Please contact the app developer.

The crm.objects.deals.write scope is the culprit. I take it out and we're back to the Insufficient scopes were provided error.

Maybe you'll have more luck:

https://app.hubspot.com/oauth/authorize?client_id=0cc5f843-a3ac-4d55-aa52-f94a6f701fd6&response_type=code&redirect_uri=https%3a%2f%2fglobal.consent.azure-apim.net%2fredirect&scope=content%20reports%20social%20automation%20actions%20timeline%20business-intelligence%20oauth%20forms%20files%20hubdb%20transactional-email%20integration-sync%20tickets%20e-commerce%20accounting%20sales-email-read%20forms-uploaded-files%20communication_preferences.read_write%20crm.import%20communication_preferences.read%20communication_preferences.write%20conversations.visitor_identification.tokens.create%20files.ui_hidden.read%20media_bridge.read%20media_bridge.write%20conversations.read%20conversations.write%20collector.graphql_schema.read%20collector.graphql_query.execute%20analytics.behavioral_events.send%20account-info.security.read%20crm.export%20crm.lists.read%20crm.objects.contacts.read%20crm.objects.marketing_events.read%20crm.schemas.custom.read%20crm.objects.custom.read%20crm.schemas.contacts.read%20crm.objects.feedback_submissions.read%20crm.objects.companies.read%20crm.objects.deals.read%20crm.schemas.companies.read%20crm.schemas.deals.read%20crm.objects.owners.read%20crm.objects.quotes.read%20crm.schemas.quotes.read%20crm.objects.line_items.read%20crm.schemas.line_items.read%20crm.objects.contacts.write%20crm.objects.marketing_events.write%20crm.objects.custom.write%20crm.objects.companies.write%20crm.lists.write%20crm.schemas.companies.write%20crm.schemas.contacts.write%20crm.schemas.deals.write%20crm.objects.quotes.write%20crm.objects.line_items.write%20content%20reports%20social%20automation%20actions%20timeline%20business-intelligence%20oauth%20forms%20files%20hubdb%20transactional-email%20integration-sync%20tickets%20e-commerce%20accounting%20sales-email-read%20forms-uploaded-files%20communication_preferences.read_write%20crm.import%20communication_preferences.read%20communication_preferences.write%20conversations.visitor_identification.tokens.create%20files.ui_hidden.read%20media_bridge.read%20media_bridge.write%20conversations.read%20conversations.write%20collector.graphql_schema.read%20collector.graphql_query.execute%20analytics.behavioral_events.send%20account-info.security.read%20crm.export%20crm.objects.deals.write

**Comment by Uthoocha** [comment]
@marctre123 I'm going to leave my work around as it is right now (Using Private App), because I want to try and build on Troys amazing work, such as the small change I added to search for Contacts using email instead of Customer_ID.
Building the solution as a Custom Connector is helping me understand how this all works. and Troys video on Youtube is also a good watch. This plugin could save our company thousands of pounds, replacing the 3rd party solution being used which has half the functionality of Power Apps.

**Comment by marctre123** [comment]
@troystaylor I tried @Uthoocha approach. Works like a charm. His approach is clean as he is not using the old approach of API keys. I know it sounds confusing because the authentication method he chose is API key when creating the custom connector, but it is not the same thing. If you're familiar with Azure, you could think of it as using a SAS key (which permits granular access) instead of an Access Key (global access). 

I was aiming to use the publicly certified connector at first to profit from a maintained connector, but now I'm thinking I might have to create a custom connector anyway. The public connector must only expose the OOTB fields and not custom ones, so if I want to add my own fields like you then I must also extend it as a custom connector.

@troystaylor I think the App (Connecting Microsoft HubSpot CRM to HubSpot) the connector is trying to install in our environment is missing some scopes. For example, you have an action in your connector to update deals so it would just make sense that crm.objects.deals.write be part of your scopes. If you miss one scope in your App Install Url, it will give the message reported in this issue. So if you check the app Microsoft HubSpot CRM you created with Client ID 0cc5f843-a3ac-4d55-aa52-f94a6f701fd6, it should have an App Install Url. All the scopes specified must also be specified in your connector definition I believe. At least that's my theory. =) 

Thanks @Uthoocha, your solution combined with Troy's work will save the day for us as well.

**Comment by eduardo-smaia** [comment]
@Uthoocha , could you please give us a little more detail on how you created the Hubspot app? I'm getting the error "The redirect_uri doesn't match the client registered redirect_uri. Please contact the app developer." What should I paste in the Redirect URL field on Hubspot?

**Comment by Uthoocha** [comment]
@eduardo-smaia I used a Private App in Hubspot so I don't think it asked me for a Redirect URL.

**Comment by salaros** [comment]
wow @Uthoocha that's great. Please contact me at salaros at salaros dot com

1. I want to know more about your changes, so I could replicate them
2. I'd like to offer you some help in creating a pull request, so that your changes could be integrated into the mainstream project and other users could benefit from it

**Comment by ABzand** [comment]
This worked for me @Uthoocha! One thing to not miss is that in the test section where you enter the API Key, you need to actually include "bearer " including the space, before the actual key.

**Comment by h22221** [comment]
@Uthoocha, I know this is an old thread.... however, I am trying to use get email V3 to take contact details out of the body of an email (have it all working) and then create a contact in HubSpot. I've tried everything, and stumbled on this post at 12.30am! 
To be frank, I don't know how to make custom connectors etc, I'm new to all this as I'm a start up, but I like to try to DOY! I partially understand your instructions above. So would you please be able to walk me through what I need to do please? Any help would be greatly appreciated.
Thanks

**Comment by troystaylor** [comment]
@h22221, please try to use the HubSpot V2 connectors - those use a HubSpot private app token to connect and then you don’t have to worry about the scopes.

**Comment by h22221** [comment]
@troystaylor thanks, I'll have a look, this issue is that I don't have premium power automate.. .

**Comment by h22221** [comment]
@troystaylor & @Uthoocha - i have sorted this! in the end I got power automate to get email (which is sent from forminator on worpress when a user registers on my site), PA then sends the contact details to a sharepoint list, azure looks at the list for new entries, then http's to hubspot....it works, and hopefully free as you get 4k billable actions/triggers with azure. Thought I'd update you as may be another work around without a premium power automate subscription!
